File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/MetricsFactory.java
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class MetricsFactory {

    private final List<Class<? extends Metric>> metricClasses;

    public MetricsFactory() {
        this.metricClasses = createAllMetrics().stream()
                .map(Metric::getClass)
                .collect(Collectors.toList());
    }

    public MetricsFactory(List<String> metricNames) {
        this.metricClasses = createAllMetrics().stream()
                .filter(m -> metricNames.contains(m.metricName()))
                .map(Metric::getClass)
                .collect(Collectors.toList());

    }

    private static Metric createMetric(Class<? extends Metric> clazz) {
        try {
            return clazz.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            throw new IllegalArgumentException("metric " + clazz + " not found.");
        }
    }

    private List<Metric> createAllMetrics() {
        return Arrays.asList(
                new AbsoluteCodeChurn(),
                new LinesOfCode(),
                new NumberOfAuthors(),
                new NumberOfOccurencesInCommits(),
                new RangeOfWeeksWithCommits(),
                new RelativeCodeChurn(),
                new SuccessiveWeeksWithCommits(),
                new WeeksWithCommits(),
                new HighlyCoupledFiles(),
                new MedianCoupledFiles(),
                new AbsoluteCoupledChurn(),
                new AverageCodeChurnPerCommit(),
                new AddQuotDel()
        );
    }

    public List<Metric> createMetrics() {
        return metricClasses.stream()
                .map(MetricsFactory::createMetric)
                .collect(Collectors.toList());
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/MetricsFactory.kt
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics

import java.util.*

class MetricsFactory {

    private val metricClasses: List<Class<out Metric>>

    constructor() {
        this.metricClasses = createAllMetrics()
                .map { it.javaClass }
    }

    constructor(metricNames: List<String>) {
        this.metricClasses = createAllMetrics()
                .filter { m -> metricNames.contains(m.metricName()) }
                .map { it.javaClass }
    }

    private fun createMetric(clazz: Class<out Metric>): Metric {
        try {
            return clazz.newInstance()
        } catch (e: InstantiationException) {
            throw IllegalArgumentException("metric $clazz not found.")
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException("metric $clazz not found.")
        }
    }

    private fun createAllMetrics(): List<Metric> {
        return Arrays.asList(
                AbsoluteCodeChurn(),
                AddedLines(),
                DeletedLines(),
                NumberOfAuthors(),
                NumberOfOccurencesInCommits(),
                RangeOfWeeksWithCommits(),
                SuccessiveWeeksWithCommits(),
                WeeksWithCommits(),
                HighlyCoupledFiles(),
                MedianCoupledFiles(),
                AbsoluteCoupledChurn(),
                AverageCodeChurnPerCommit(),
                NumberOfRenames(),
                AgeInWeeks()
        )
    }

    fun createMetrics(): List<Metric> {
        return metricClasses
                .map { createMetric(it) }
    }
}
